pkg/utils/utils.go

```diff
@@ -185,6 +185,21 @@ func GetSriovVFcapacity(pf string) int {
 	return numvfs
 }
 
+// GetDevNode returns the numa node of a PCI device, -1 if none is specified or error.
```

Comment:
> Returning '-1' on error doesn't look like a good idea. This is a normal case when kernel was built without NUMA support and in this case where will be no 'numa_node' knob or it will always return '-1' as a result. I think that it's better to return 0 in these cases. Otherwise you'll not be able to ever start a pod with 'strict' policy on a node without NUMA support (like inside the VM).

I agree, we need to cognizant of non-numa systems. Let's verify how the Topology manager handles the non-numa scenario and what do they expect from device plugin in response from such system.

Reply:
kubernetes uses cadvisor to get topology information. cadvisor parses /proc/cpuinfo for 'physical id' and counts number of nodes this way. So, for non-NUMA systems there be single node with id 0. For arches without 'physical id' in cpuinfo it parses /sys/bus/cpu/devices for 'nodeX' files. On failure it reports node id 0.
ref: https://github.com/google/cadvisor/blob/d200c9f558b0a8051d728fd8bb74d0469c6a36e1/machine/machine.go#L194

Reply:
I'll go back and ping the folks working on the k8s side of things as to what the expectations are for the case where the kernel returns "-1" (due to the device not having numa affinity) or the host kernel was built without numa awareness.

I actually had intentionally used -1 on error to match what the kernel returns, but maybe they want nil instead.  Returning 0 is definitely not the right thing to do, as in a VM it's possible to have multiple virtual NUMA nodes but devices with numa_node=-1.

Reply:
Okay, based on the reply at https://github.com/kubernetes/kubernetes/pull/80570#issuecomment-523125339 the expectation is that if we cannot determine a NUMA node then we should simply not set the field.  I'll update the code to reflect that.

Given this, I think it actually makes sense to preserve the "-1" in GetDevNode() and then check for that to decide whether or not to set the "topology" field at all.

Reply:
Please let me know if you see any issues with this approach.

Reply:
setting the field to -1 or not setting it is just semantics as PB definition has this field defined with a default of -1.

it would be useful to state what -1 means to devicemanager in PB API definition & relevant documentation.

Reply:
To expand on what happens in the `devicemanager`/`TopologyManager` in this case:

From the `devicemanager`'s perspective, having even one device with `Topology.Node` unset, will result in it returning a `nil` for the `[]TopologyHint` for that device type.

When the `TopologyManager` sees a `nil` `[]TopologyHint` for a given resource type, it treats that as a "don't care" alignment for that resource type and generates a single hint with a full socket mask for that resource type (indicating allocations can occur from any socket for that resource type). This will then be combined with hints from other resource types to generate a "merged" hint with a common alignment amongst all resource types.

When the `strict` policy (recently renamed as `restricted`) performs its merge strategy, it will only admit the pod if resource alignment on a _single_ NUMA node can be satisfied (i.e. if even one resource from any resource type needs to be pulled from a second NUMA node, the pod will not be admitted).

So in the case being discussed here (i.e. of a VM only seeing a single NUMA node and returning -1 as its Node ID), we should still be OK (even with the `restricted` policy).

The `devicemanager` will return `nil` for its hints on all resource types, the `restricted` policy will turn this into a hint with a filled socketmask with only one bit set (because there is only one socket on the machine) and it will admit the pod because alignment can (and by construction must be) satisfied on that single socket in the mask.

Reply:
@klueska Thanks for explaining the behavior. I think from device plugin point of view it only make sense to add `TopologyInfo` if we able get valid NUMA info from the platform. Otherwise we should leave this field to be `nil` and let device manager decide how this information is communicated toTopologyManager.

Reply:
Agreed

Reply:
> From the `devicemanager`'s perspective, having even one device with `Topology.Node` unset, will result in it returning a `nil` for the `[]TopologyHint` for that device type.
> 
> When the `TopologyManager` sees a `nil` `[]TopologyHint` for a given resource type, it treats that as a "don't care" alignment for that resource type and generates a single hint with a full socket mask for that resource type (indicating allocations can occur from any socket for that resource type). This will then be combined with hints from other resource types to generate a "merged" hint with a common alignment amongst all resource types.

So, in case we are in a NUMA system, Numa affinity is required and we failed to get the NodeID for whatever reason for a **single** device in a (SR-IOV) resource then it may lead to a TopologyHint that may not be satisfiable by devicemanager for that specific resource ?

Reply:
As it's implemented right now, if any single device for a given device type does not have the the `Topology` field set (even if all of the other devices of the same type do have it set), then the `devicemanager` will return `nil` for its list of hints for that device type (resulting in a "don't care" alignment for that device type).

Reply:
An alternative would be to set the socketmask to match the devices that do have the `Topology` field set and just ignore the devices that don't have it set.

In the case of the `restricted` policy, this would result in only admitting pods that matched alignment against the devices it could read the topology from (i.e. the device without `Topology` would never be available for allocation). In the case of the `best-effort` policy, the device without `Topology` would end up being a sort of "last resort" device assuming no other alignment could be satisfied.

Are these semantics preferred to the existing ones?

Reply:
That's what i had in mind.

Reply:
Upon reflection, I think these "newer" semantics are more correct than the old ones (even if they aren't perfect).

In the existing implementation, both `restricted´ and `best-effort` policies have the ability to admit pods that may not have true alignment (because falling back to "don't care" when no `Topology` field is set, essentially makes devices of that type appear to be aligned on every socket and thus allocatable). This goes against the entire purpose of the `restricted` policy and would likely lead to much confusion down the road.

I am going to update https://github.com/kubernetes/kubernetes/pull/80570 to reflect the updated semantics.

Reply:
If a host kernel is built without CONFIG_NUMA I think the `numa_node` file for all devices on that host will simply be missing and so there would be no `Topology` field.

If an pod really cares about NUMA-affinity, it would kind of defeat the purpose to put it on a host that is not NUMA-aware.

As such, I think it would make sense that devices without `Topology` wouldn't match for the `strict` policy.